predicators/spot_utils/perception/object_detection.py

```diff
@@ -0,0 +1,513 @@
+        artifacts["april"].update(img_artifacts)
+
+    # There IS batching over images here for efficiency.
+    language_detections, language_artifacts = detect_objects_from_language(
+        language_object_ids, rgbds)
+    detections.update(language_detections)
+    artifacts["language"] = language_artifacts
+
+    return detections, artifacts
+
+
+def detect_objects_from_april_tags(
+    object_ids: Collection[AprilTagObjectDetectionID],
+    rgbd: RGBDImageWithContext,
+    fiducial_size: float = CFG.spot_fiducial_size,
+) -> Tuple[Dict[ObjectDetectionID, math_helpers.SE3Pose], Dict[str, Any]]:
+    """Detect an object pose from an april tag.
+
+    The rotation is currently not detected (set to default).
+    """
+    tag_num_to_object_id = {t.april_tag_number: t for t in object_ids}
+
+    # Convert the RGB image to grayscale.
+    image_grey = cv2.cvtColor(rgbd.rgb, cv2.COLOR_RGB2GRAY)
+
+    # Create apriltag detector and get all apriltag locations.
+    options = apriltag.DetectorOptions(families="tag36h11")
+    options.refine_pose = 1
+    detector = apriltag.Detector(options)
+    apriltag_detections = detector.detect(image_grey)
+
+    detections: Dict[ObjectDetectionID, math_helpers.SE3Pose] = {}
+    artifacts: Dict[str, Any] = {}
+
+    # For every detection, find pose in world frame.
+    for apriltag_detection in apriltag_detections:
+        # Only include requested tags.
+        if apriltag_detection.tag_id not in tag_num_to_object_id:
+            continue
+        obj_id = tag_num_to_object_id[apriltag_detection.tag_id]
+
+        # Save the detection for external analysis.
+        artifact_id = f"apriltag_{rgbd.camera_name}_{obj_id.april_tag_number}"
+        artifacts[artifact_id] = apriltag_detection
+
+        # Get the pose from the apriltag library.
+        pose = detector.detection_pose(
+            apriltag_detection,
+            (rgbd.intrinsics.focal_length.x, rgbd.intrinsics.focal_length.y,
+             rgbd.intrinsics.principal_point.x,
+             rgbd.intrinsics.principal_point.y), fiducial_size)[0]
+        tx, ty, tz, tw = pose[:, -1]
+        assert np.isclose(tw, 1.0)
+
+        # Detection is in meters, we want mm.
+        camera_frame_pose = math_helpers.SE3Pose(
+            x=float(tx) / 1000.0,
+            y=float(ty) / 1000.0,
+            z=float(tz) / 1000.0,
+            rot=math_helpers.Quat(),
+        )
+
+        # Apply transforms.
+        world_frame_pose = rgbd.world_tform_camera * camera_frame_pose
+        world_frame_pose = obj_id.offset_transform * world_frame_pose
+
+        # Save in detections.
+        detections[obj_id] = world_frame_pose
+
+    return detections, artifacts
+
+
+def detect_objects_from_language(
+    object_ids: Collection[LanguageObjectDetectionID],
+    rgbds: Dict[str, RGBDImageWithContext],
+) -> Tuple[Dict[ObjectDetectionID, math_helpers.SE3Pose], Dict[str, Any]]:
+    """Detect an object pose using a vision-language model."""
```

Comment:
ditto - mention that the second return value is artifacts, and ideally mention what those are